ArduCopter/mode_guided.cpp

```diff
@@ -493,15 +581,107 @@ void ModeGuided::vel_control_run()
     // set velocity to zero and stop rotating if no updates received for 3 seconds
     uint32_t tnow = millis();
     if (tnow - vel_update_time_ms > GUIDED_POSVEL_TIMEOUT_MS) {
-        if (!pos_control->get_vel_desired_cms().is_zero()) {
-            set_desired_velocity_with_accel_and_fence_limits(Vector3f(0.0f, 0.0f, 0.0f));
-        }
+        guided_accel_target_cmss.zero();
         if (auto_yaw.mode() == AUTO_YAW_RATE) {
             auto_yaw.set_rate(0.0f);
         }
+    }
+
+    // update position controller with new target
+    pos_control->input_accel_xy(guided_accel_target_cmss);
+    if (!stabilizing_vel_xy()) {
+        // set position and velocity errors to zero
+        pos_control->stop_vel_xy_stabilisation();
+    } else if (!stabilizing_pos_xy()) {
+        // set position errors to zero
+        pos_control->stop_pos_xy_stabilisation();
+    }
+    pos_control->input_vel_accel_z(guided_vel_target_cms, guided_accel_target_cmss, false);
+
+    // call velocity controller which includes z axis controller
+    pos_control->update_xy_controller();
+    pos_control->update_z_controller();
+
+    // call attitude controller
+    if (auto_yaw.mode() == AUTO_YAW_HOLD) {
+        // roll & pitch from waypoint controller, yaw rate from pilot
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), target_yaw_rate);
+    } else if (auto_yaw.mode() == AUTO_YAW_RATE) {
+        // roll & pitch from velocity controller, yaw rate from mavlink command or mission item
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), auto_yaw.rate_cds());
     } else {
-        set_desired_velocity_with_accel_and_fence_limits(guided_vel_target_cms);
+        // roll, pitch from waypoint controller, yaw heading from GCS or auto_heading()
+        attitude_control->input_thrust_vector_heading(pos_control->get_thrust_vector(), auto_yaw.yaw());
     }
+}
+
+// velaccel_control_run - runs the guided velocity controller
```

Comment:
comment should probably be, "runs the guided velocity and acceleration controller"